File: JMStaticContentTableViewController/JMStaticContentTableViewController.m
#import "JMStaticContentTableViewController.h"
#import <QuartzCore/QuartzCore.h>

#if __IPHONE_OS_VERSION_MAX_ALLOWED < 60000
#define NSTextAlignmentCenter       UITextAlignmentCenter
#endif


@implementation JMStaticContentTableViewController

@synthesize staticContentSections = _staticContentSections;

@synthesize headerText = _headerText;
@synthesize footerText = _footerText;

#pragma mark - View lifecycle

- (void) viewDidLoad {
    [super viewDidLoad];

	[self removeAllSections];
}
- (void) viewDidUnload {
    [super viewDidUnload];
}

- (void) didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

#pragma mark - Table view data source

- (NSInteger) numberOfSectionsInTableView:(UITableView *)tableView {
    return self.staticContentSections.count;
}
- (NSInteger) tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
	JMStaticContentTableViewSection *sectionContent = [self.staticContentSections objectAtIndex:section];

    return sectionContent.staticContentCells.count;
}

- (CGFloat) tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
	JMStaticContentTableViewSection *sectionContent = [self.staticContentSections objectAtIndex:indexPath.section];
	JMStaticContentTableViewCell *cellContent = [sectionContent.staticContentCells objectAtIndex:indexPath.row];

	return cellContent.cellHeight;
}

- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section {
    JMStaticContentTableViewSection *sectionContent = [self.staticContentSections objectAtIndex:section];
    UIView *headerView = sectionContent.headerView;
	if (headerView) {
        return CGRectGetHeight(headerView.frame);
    }
    
    return UITableViewAutomaticDimension;
}

- (CGFloat)tableView:(UITableView *)tableView heightForFooterInSection:(NSInteger)section {
    JMStaticContentTableViewSection *sectionContent = [self.staticContentSections objectAtIndex:section];
    UIView *footerView = sectionContent.headerView;
	if (footerView) {
        return CGRectGetHeight(footerView.frame);
    }
    
    return UITableViewAutomaticDimension;
}

- (NSString *) tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
	JMStaticContentTableViewSection *sectionContent = [self.staticContentSections objectAtIndex:section];
	return sectionContent.headerTitle;
}
- (NSString *) tableView:(UITableView *)tableView titleForFooterInSection:(NSInteger)section {
	JMStaticContentTableViewSection *sectionContent = [self.staticContentSections objectAtIndex:section];
	return sectionContent.footerTitle;
}

- (UIView *)tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger)section {
	JMStaticContentTableViewSection *sectionContent = [self.staticContentSections objectAtIndex:section];
	return sectionContent.headerView;
}

- (UIView *)tableView:(UITableView *)tableView viewForFooterInSection:(NSInteger)section {
	JMStaticContentTableViewSection *sectionContent = [self.staticContentSections objectAtIndex:section];
	return sectionContent.footerView;
}

- (UITableViewCell *) tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
	JMStaticContentTableViewSection *sectionContent = [self.staticContentSections objectAtIndex:indexPath.section];
	JMStaticContentTableViewCell *cellContent = [sectionContent.staticContentCells objectAtIndex:indexPath.row];

    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:cellContent.reuseIdentifier];

    if (cell == nil) {
		cell = [[[cellContent tableViewCellSubclass] alloc] initWithStyle:cellContent.cellStyle reuseIdentifier:cellContent.reuseIdentifier];
    }

	cell.imageView.image = nil;
	cell.accessoryView = nil;
	cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;

	cellContent.configureBlock(cellContent, cell, indexPath);

    return cell;
}

#pragma mark - Table view delegate

- (UITableViewCellEditingStyle) tableView:(UITableView *)tableView editingStyleForRowAtIndexPath:(NSIndexPath *)indexPath {
	JMStaticContentTableViewSection *sectionContent = [self.staticContentSections objectAtIndex:indexPath.section];
	JMStaticContentTableViewCell *cellContent = [sectionContent.staticContentCells objectAtIndex:indexPath.row];

    return cellContent.editingStyle;
}

- (BOOL) tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
	JMStaticContentTableViewSection *sectionContent = [self.staticContentSections objectAtIndex:indexPath.section];
	JMStaticContentTableViewCell *cellContent = [sectionContent.staticContentCells objectAtIndex:indexPath.row];

    return cellContent.editable;
}

- (BOOL) tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)indexPath {
	JMStaticContentTableViewSection *sectionContent = [self.staticContentSections objectAtIndex:indexPath.section];
	JMStaticContentTableViewCell *cellContent = [sectionContent.staticContentCells objectAtIndex:indexPath.row];

    return cellContent.moveable;
}

- (void) tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
	if(!tableView.editing && !tableView.allowsMultipleSelection) [tableView deselectRowAtIndexPath:indexPath animated:YES];
	if(tableView.editing && !tableView.allowsMultipleSelectionDuringEditing) [tableView deselectRowAtIndexPath:indexPath animated:YES];

	JMStaticContentTableViewSection *sectionContent = [self.staticContentSections objectAtIndex:indexPath.section];
	JMStaticContentTableViewCell *cellContent = [sectionContent.staticContentCells objectAtIndex:indexPath.row];

	if(cellContent.whenSelectedBlock) {
		cellContent.whenSelectedBlock(indexPath);
	}
}
- (void) scrollViewDidEndDragging:(UIScrollView *)scrollView willDecelerate:(BOOL)decelerate {
	[self.view endEditing:YES];
}

#pragma mark - Static Content

- (void) _updateSectionIndexes {
    NSInteger updatedIndex = 0;
    for(JMStaticContentTableViewSection *section in self.staticContentSections) {
        section.sectionIndex = updatedIndex;
        updatedIndex++;
    }
}

- (void) removeAllSections {
	if(self.staticContentSections) {
		self.staticContentSections = nil;
		self.staticContentSections = [NSArray array];
	}
}

- (void) addSection:(JMStaticContentTableViewControllerAddSectionBlock)b {
	if(!self.staticContentSections) self.staticContentSections = [NSArray array];
	
	JMStaticContentTableViewSection *section = [[JMStaticContentTableViewSection alloc] init];
	section.tableView = self.tableView;
    section.sectionIndex = [self.staticContentSections count];

	b(section, section.sectionIndex);

	self.staticContentSections = [self.staticContentSections arrayByAddingObject:section];

    [self _updateSectionIndexes];
}

- (void) insertSection:(JMStaticContentTableViewControllerAddSectionBlock)b atIndex:(NSUInteger)sectionIndex {
	[self insertSection:b atIndex:sectionIndex animated:YES];
}
- (void) insertSection:(JMStaticContentTableViewControllerAddSectionBlock)b atIndex:(NSUInteger)sectionIndex animated:(BOOL)animated {
	[self insertSection:b atIndex:sectionIndex animated:YES updateView:YES];
}

- (void) insertSection:(JMStaticContentTableViewControllerAddSectionBlock)b atIndex:(NSUInteger)sectionIndex animated:(BOOL)animated updateView:(BOOL)updateView {
	if(!self.staticContentSections) self.staticContentSections = [NSArray array];

	NSMutableArray *mutableSections = [self.staticContentSections mutableCopy];

	JMStaticContentTableViewSection *section = [[JMStaticContentTableViewSection alloc] init];

	b(section, sectionIndex);

	[mutableSections insertObject:section atIndex:sectionIndex];

	self.staticContentSections = [NSArray arrayWithArray:mutableSections];

    [self _updateSectionIndexes];

    if (updateView) {
        if(animated) {
            [self.tableView insertSections:[NSIndexSet indexSetWithIndex:sectionIndex] withRowAnimation:UITableViewRowAnimationAutomatic];
        } else {
            [self.tableView reloadData];
        }
    }
}

- (void) removeSectionAtIndex:(NSUInteger)sectionIndex {
	[self removeSectionAtIndex:sectionIndex animated:YES];
}
- (void) removeSectionAtIndex:(NSUInteger)sectionIndex animated:(BOOL)animated {
	NSMutableArray *sections = [self.staticContentSections mutableCopy];

	[sections removeObjectAtIndex:sectionIndex];

	self.staticContentSections = [NSArray arrayWithArray:sections];

    [self _updateSectionIndexes];

	if(animated) {
		[self.tableView beginUpdates];

		[self.tableView deleteSections:[NSIndexSet indexSetWithIndex:sectionIndex] withRowAnimation:UITableViewRowAnimationAutomatic];

		[self.tableView endUpdates];
	} else {
		[self.tableView reloadData];
	}
}

- (void) reloadSectionAtIndex:(NSUInteger)sectionIndex {
	[self reloadSectionAtIndex:sectionIndex animated:YES];
}
- (void) reloadSectionAtIndex:(NSUInteger)sectionIndex animated:(BOOL)animated {
    [self.tableView reloadSections:[NSIndexSet indexSetWithIndex:sectionIndex] withRowAnimation:animated ? UITableViewRowAnimationAutomatic : UITableViewRowAnimationNone];
}

- (JMStaticContentTableViewSection *) sectionAtIndex:(NSUInteger)sectionIndex {
	return [self.staticContentSections objectAtIndex:sectionIndex];
}

- (void) insertCell:(JMStaticContentTableViewCellBlock)configurationBlock 
		atIndexPath:(NSIndexPath *)indexPath 
		   animated:(BOOL)animated {

	[self insertCell:configurationBlock
		whenSelected:nil
		 atIndexPath:indexPath
			animated:YES];
}

- (void) insertCell:(JMStaticContentTableViewCellBlock)configurationBlock
	   whenSelected:(JMStaticContentTableViewCellWhenSelectedBlock)whenSelectedBlock
		atIndexPath:(NSIndexPath *)indexPath 
		   animated:(BOOL)animated {

	JMStaticContentTableViewSection *section = [self sectionAtIndex:indexPath.section];

	[section insertCell:configurationBlock 
		   whenSelected:whenSelectedBlock 
			atIndexPath:indexPath
			   animated:animated];
}

#pragma mark - Headers & Footers

- (void) setHeaderText:(NSString *)headerTextValue {
    _headerText = headerTextValue;
    
	if(!headerTextValue) {
		if([self isViewLoaded]) {
			self.tableView.tableFooterView = nil;
		}
		
		return;
	}
	
	if([self isViewLoaded]) {
		UIView *headerLabelContainerView = [[UIView alloc] initWithFrame:CGRectMake(0.0, 0.0, 320.0, 0.0)];
		UILabel *headerLabel = [[UILabel alloc] initWithFrame:CGRectMake(10.0, 10.0, 280.0, 0.0)];
        
		headerLabel.backgroundColor = [UIColor clearColor];
		headerLabel.font = [UIFont systemFontOfSize:15.0];
		headerLabel.textColor = [UIColor colorWithRed:61.0/255.0 green:77.0/255.0 blue:99.0/255.0 alpha:1.0];
		headerLabel.shadowColor = [UIColor colorWithWhite:1.0 alpha:0.65];
		headerLabel.shadowOffset = CGSizeMake(0.0, 1.0);
		headerLabel.textAlignment = NSTextAlignmentCenter;
		headerLabel.numberOfLines = 0;
        
		headerLabel.autoresizingMask = UIViewAutoresizingFlexibleWidth;
        
		headerLabel.text = self.headerText;
        
		[headerLabel sizeToFit];
        
		CGRect headerLabelContainerViewRect = headerLabelContainerView.frame;
		headerLabelContainerViewRect.size.height = headerLabel.frame.size.height + 10.0;
		headerLabel.frame = headerLabelContainerViewRect;
        
		[headerLabelContainerView addSubview:headerLabel];
        
		CGRect headerLabelFrame = headerLabel.frame;
		headerLabelFrame.size.width = 280.0;
		headerLabelFrame.origin.x = 20.0;
		headerLabelFrame.origin.y = 10.0;	
        headerLabelFrame.size.height += 10.0;
		headerLabel.frame = headerLabelFrame;
        
		CGRect containerFrame = headerLabelContainerView.frame;
		containerFrame.size.height = headerLabel.frame.size.height + 10.0;
		headerLabelContainerView.frame = containerFrame;
        
		self.tableView.tableHeaderView = headerLabelContainerView;
	}
}

- (void) setFooterText:(NSString *)footerTextValue {
    _footerText = footerTextValue;

	if(!footerTextValue) {
		if([self isViewLoaded]) {
			self.tableView.tableFooterView = nil;
		}
		
		return;
	}
	
	if([self isViewLoaded]) {
		UIView *footerLabelContainerView = [[UIView alloc] initWithFrame:CGRectMake(0.0, 0.0, 320.0, 0.0)];
		UILabel *footerLabel = [[UILabel alloc] initWithFrame:CGRectMake(10.0, 10.0, 280.0, 0.0)];

		footerLabel.backgroundColor = [UIColor clearColor];
		footerLabel.font = [UIFont systemFontOfSize:15.0];
		footerLabel.textColor = [UIColor colorWithRed:61.0/255.0 green:77.0/255.0 blue:99.0/255.0 alpha:1.0];
		footerLabel.shadowColor = [UIColor colorWithWhite:1.0 alpha:0.65];
		footerLabel.shadowOffset = CGSizeMake(0.0, 1.0);
		footerLabel.textAlignment = NSTextAlignmentCenter;
		footerLabel.numberOfLines = 0;

		footerLabel.autoresizingMask = UIViewAutoresizingFlexibleWidth;

		footerLabel.text = self.footerText;

		[footerLabel sizeToFit];

		CGRect footerLabelContainerViewRect = footerLabelContainerView.frame;
		footerLabelContainerViewRect.size.height = footerLabel.frame.size.height + 10.0;
		footerLabel.frame = footerLabelContainerViewRect;

		[footerLabelContainerView addSubview:footerLabel];

		CGRect footerLabelFrame = footerLabel.frame;
		footerLabelFrame.size.width = 260.0;
		footerLabelFrame.origin.x = 30.0;
		footerLabelFrame.origin.y = 0.0;	
		footerLabel.frame = footerLabelFrame;

		CGRect containerFrame = footerLabelContainerView.frame;
		containerFrame.size.height = footerLabel.frame.size.height + 10.0;
		footerLabelContainerView.frame = containerFrame;

		self.tableView.tableFooterView = footerLabelContainerView;
	}
}

@end

File: Sources/JMStaticContentTableViewController/JMStaticContentTableViewController.swift
//
//  JMStaticContentTableViewController.swift
//  JMStaticContentTableViewController
//
//  Created by Jake Marsh on 10/8/11.
//  Rewritten in Swift on 12/31/24.
//  Copyright 2011-2025 Jake Marsh. All rights reserved.
//

#if canImport(UIKit) && !os(watchOS)
import UIKit

/// A table view controller for static content like settings screens.
///
/// `JMStaticContentTableViewController` provides a simple, declarative way to create
/// static table views with sections and cells, similar to iOS Settings.
///
/// ## Basic Usage
///
/// ```swift
/// class SettingsViewController: JMStaticContentTableViewController {
///     override func viewDidLoad() {
///         super.viewDidLoad()
///
///         addSection { section, _ in
///             section.headerTitle = "Account"
///
///             section.addCell(.text("Profile", onSelect: { _ in
///                 // Navigate to profile
///             }))
///
///             section.addCell(.toggle("Notifications", isOn: true) { isOn in
///                 // Handle toggle
///             })
///         }
///     }
/// }
/// ```
open class JMStaticContentTableViewController: UITableViewController {

    // MARK: - Properties

    /// All sections in the table view.
    public private(set) var sections: [JMStaticContentSection] = []

    /// Text displayed in the table header.
    public var headerText: String? {
        didSet { updateHeaderView() }
    }

    /// Text displayed in the table footer.
    public var footerText: String? {
        didSet { updateFooterView() }
    }

    // MARK: - Lifecycle

    open override func viewDidLoad() {
        super.viewDidLoad()
        removeAllSections()
    }

    // MARK: - Section Management

    /// Adds a new section to the table view.
    /// - Parameter block: Configuration block for the section.
    public func addSection(_ block: (JMStaticContentSection, Int) -> Void) {
        let section = JMStaticContentSection()
        section.tableView = tableView
        section.sectionIndex = sections.count

        block(section, section.sectionIndex)

        sections.append(section)
        updateSectionIndexes()
    }

    /// Inserts a section at the specified index.
    /// - Parameters:
    ///   - block: Configuration block for the section.
    ///   - index: The index to insert at.
    ///   - animated: Whether to animate the insertion.
    public func insertSection(
        _ block: (JMStaticContentSection, Int) -> Void,
        at index: Int,
        animated: Bool = true
    ) {
        let section = JMStaticContentSection()
        section.tableView = tableView

        block(section, index)

        sections.insert(section, at: index)
        updateSectionIndexes()

        if animated {
            tableView.insertSections(IndexSet(integer: index), with: .automatic)
        } else {
            tableView.reloadData()
        }
    }

    /// Removes all sections from the table view.
    ///
    /// This method properly synchronizes the data source with the table view to prevent crashes.
    public func removeAllSections() {
        sections.removeAll()
        // FIX for issue #4: Always reload data after removing sections
        tableView.reloadData()
    }

    /// Removes the section at the specified index.
    /// - Parameters:
    ///   - index: The index of the section to remove.
    ///   - animated: Whether to animate the removal.
    public func removeSection(at index: Int, animated: Bool = true) {
        guard index < sections.count else { return }

        sections.remove(at: index)
        updateSectionIndexes()

        if animated {
            tableView.beginUpdates()
            tableView.deleteSections(IndexSet(integer: index), with: .automatic)
            tableView.endUpdates()
        } else {
            tableView.reloadData()
        }
    }

    /// Reloads the section at the specified index.
    /// - Parameters:
    ///   - index: The index of the section to reload.
    ///   - animated: Whether to animate the reload.
    public func reloadSection(at index: Int, animated: Bool = true) {
        guard index < sections.count else { return }

        let animation: UITableView.RowAnimation = animated ? .automatic : .none
        tableView.reloadSections(IndexSet(integer: index), with: animation)
    }

    /// Returns the section at the specified index.
    /// - Parameter index: The section index.
    /// - Returns: The section at the index, or nil if out of bounds.
    public func section(at index: Int) -> JMStaticContentSection? {
        guard index < sections.count else { return nil }
        return sections[index]
    }

    /// Inserts a cell at the specified index path.
    /// - Parameters:
    ///   - cell: The cell configuration to insert.
    ///   - indexPath: The index path for the new cell.
    ///   - animated: Whether to animate the insertion.
    public func insertCell(_ cell: JMStaticContentCell, at indexPath: IndexPath, animated: Bool = true) {
        guard indexPath.section < sections.count else { return }

        let section = sections[indexPath.section]
        section.insertCell(cell, at: indexPath.row, animated: animated)
    }

    // MARK: - Private Methods

    private func updateSectionIndexes() {
        for (index, section) in sections.enumerated() {
            section.sectionIndex = index
        }
    }

    private func updateHeaderView() {
        guard isViewLoaded else { return }

        guard let text = headerText, !text.isEmpty else {
            tableView.tableHeaderView = nil
            return
        }

        let container = UIView()
        let label = UILabel()
        label.font = .systemFont(ofSize: 15)
        label.textColor = .secondaryLabel
        label.textAlignment = .center
        label.numberOfLines = 0
        label.text = text
        label.translatesAutoresizingMaskIntoConstraints = false

        container.addSubview(label)

        NSLayoutConstraint.activate([
            label.topAnchor.constraint(equalTo: container.topAnchor, constant: 16),
            label.leadingAnchor.constraint(equalTo: container.leadingAnchor, constant: 20),
            label.trailingAnchor.constraint(equalTo: container.trailingAnchor, constant: -20),
            label.bottomAnchor.constraint(equalTo: container.bottomAnchor, constant: -8)
        ])

        let size = container.systemLayoutSizeFitting(
            CGSize(width: tableView.bounds.width, height: UIView.layoutFittingCompressedSize.height),
            withHorizontalFittingPriority: .required,
            verticalFittingPriority: .fittingSizeLevel
        )
        container.frame = CGRect(origin: .zero, size: size)

        tableView.tableHeaderView = container
    }

    private func updateFooterView() {
        guard isViewLoaded else { return }

        guard let text = footerText, !text.isEmpty else {
            tableView.tableFooterView = nil
            return
        }

        let container = UIView()
        let label = UILabel()
        label.font = .systemFont(ofSize: 13)
        label.textColor = .secondaryLabel
        label.textAlignment = .center
        label.numberOfLines = 0
        label.text = text
        label.translatesAutoresizingMaskIntoConstraints = false

        container.addSubview(label)

        NSLayoutConstraint.activate([
            label.topAnchor.constraint(equalTo: container.topAnchor, constant: 8),
            label.leadingAnchor.constraint(equalTo: container.leadingAnchor, constant: 30),
            label.trailingAnchor.constraint(equalTo: container.trailingAnchor, constant: -30),
            label.bottomAnchor.constraint(equalTo: container.bottomAnchor, constant: -16)
        ])

        let size = container.systemLayoutSizeFitting(
            CGSize(width: tableView.bounds.width, height: UIView.layoutFittingCompressedSize.height),
            withHorizontalFittingPriority: .required,
            verticalFittingPriority: .fittingSizeLevel
        )
        container.frame = CGRect(origin: .zero, size: size)

        tableView.tableFooterView = container
    }

    // MARK: - UITableViewDataSource

    open override func numberOfSections(in tableView: UITableView) -> Int {
        return sections.count
    }

    open override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return sections[section].cells.count
    }

    open override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let section = sections[indexPath.section]
        let cellConfig = section.cells[indexPath.row]

        var cell = tableView.dequeueReusableCell(withIdentifier: cellConfig.reuseIdentifier)

        if cell == nil {
            cell = cellConfig.tableViewCellSubclass.init(style: cellConfig.cellStyle, reuseIdentifier: cellConfig.reuseIdentifier)
        }

        guard let cell else {
            return UITableViewCell()
        }

        // Reset cell state
        cell.imageView?.image = nil
        cell.accessoryView = nil
        cell.accessoryType = .disclosureIndicator

        // Configure the cell
        cellConfig.configureBlock?(cellConfig, cell, indexPath)

        return cell
    }

    open override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        return sections[section].headerTitle
    }

    open override func tableView(_ tableView: UITableView, titleForFooterInSection section: Int) -> String? {
        return sections[section].footerTitle
    }

    open override func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
        return sections[section].headerView
    }

    open override func tableView(_ tableView: UITableView, viewForFooterInSection section: Int) -> UIView? {
        return sections[section].footerView
    }

    // MARK: - UITableViewDelegate

    open override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        let cellConfig = sections[indexPath.section].cells[indexPath.row]

        if cellConfig.cellHeight == UITableView.automaticDimension {
            if tableView.rowHeight == UITableView.automaticDimension {
                return UITableView.automaticDimension
            }
            return tableView.rowHeight
        }

        return cellConfig.cellHeight
    }

    open override func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        if let headerView = sections[section].headerView {
            return headerView.frame.height
        }
        return UITableView.automaticDimension
    }

    open override func tableView(_ tableView: UITableView, heightForFooterInSection section: Int) -> CGFloat {
        if let footerView = sections[section].footerView {
            return footerView.frame.height
        }
        return UITableView.automaticDimension
    }

    open override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let cellConfig = sections[indexPath.section].cells[indexPath.row]
        cellConfig.whenSelectedBlock?(indexPath)
    }

    open override func tableView(_ tableView: UITableView, editingStyleForRowAt indexPath: IndexPath) -> UITableViewCell.EditingStyle {
        return sections[indexPath.section].cells[indexPath.row].editingStyle
    }

    open override func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
        return sections[indexPath.section].cells[indexPath.row].isEditable
    }

    open override func tableView(_ tableView: UITableView, canMoveRowAt indexPath: IndexPath) -> Bool {
        return sections[indexPath.section].cells[indexPath.row].isMoveable
    }

    // MARK: - UIScrollViewDelegate

    open override func scrollViewDidEndDragging(_ scrollView: UIScrollView, willDecelerate decelerate: Bool) {
        view.endEditing(true)
    }
}
#endif
